bot/cogs/tags.py

```diff
@@ -27,6 +26,44 @@ def __init__(self, bot: Bot):
         self.bot = bot
         self.tag_cooldowns = {}
 
+        self._cache = {}
+        self._last_fetch = None
+
+    async def _get_tags(self, is_forced: bool = False) -> None:
+        """Getting all tags."""
+        # Refresh only when there's a more than 5m gap from last call.
+        if is_forced or not self._last_fetch or time.time() - self._last_fetch > 5 * 60:
+            tags = await self.bot.api_client.get('bot/tags')
+            self._cache = {tag['title'].lower(): tag for tag in tags}
+
+    @staticmethod
+    def _fuzzy_search(search: str, target: str) -> bool:
+        found = 0
+        index = 0
+        _search = search.lower().replace(' ', '')
+        _target = target.lower().replace(' ', '')
+        for letter in _search:
+            index = _target.find(letter, index)
+            if index == -1:
+                break
+            found += index > 0
+        return found / len(_search) * 100
+
+    def _get_suggestions(self, tag_name: str, score: int) -> list:
+        return sorted(
+            (tag for tag in self._cache.values() if Tags._fuzzy_search(tag_name, tag['title']) >= score),
+            key=lambda tag: Tags._fuzzy_search(tag_name, tag['title']),
+            reverse=True
+        )
+
+    async def _get_tag(self, tag_name: str) -> list:
+        """Get a specific tag."""
+        await self._get_tags()
+        found = [self._cache.get(tag_name.lower(), None)]
+        if not found[0]:
+            return self._get_suggestions(tag_name, 100) or self._get_suggestions(tag_name, 80)
```

Comment:
This could end up doing redundant work because the first suggestions call discards tags which would be in the alternate threshold of 80. What it instead returned something like:

```py
{
    100: ["tag name 1", "another tag"],
    80: ["yet another tag"]
}
```

Reply:
Thank you for pointing that out, I've added a list of thresholds as an argument for `self._get_suggestions()` so it can go through this list and return the first one already, avoiding calling this function too many times.